sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala

```diff
@@ -270,7 +271,12 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
         .select($"eventTime".cast("long").as[Long])
 
       testStream(result, Append)(
-        StartStream(additionalConfs = Map(flagKey -> flag.toString)),
+        StartStream(additionalConfs = Map(
+          noDataBatchEnableKey -> flag.toString,
+          // set `STREAMING_NO_DATA_PROGRESS_EVENT_INTERVAL` a small value to
+          // report an `empty` progress when no data come.
+          noDataProgressIntervalKey -> "1")
```

Comment:
The way for fixing the failed UTs seems a little flaky. As your approach deletes empty progress generally, so here you set `STREAMING_NO_DATA_PROGRESS_EVENT_INTERVAL` to a small enough value to generate an empty one as expected, I think we can do better in either way:
- Add comments to explain how small `STREAMING_NO_DATA_PROGRESS_EVENT_INTERVAL` needs to be set here is enough, seems less than 10 seconds is OK?
- Delete the empty progress checking with explanation in comments.

Reply:
Let's either use the manual clock or fix the UT like the way it doesn't depend on STREAMING_NO_DATA_PROGRESS_EVENT_INTERVAL.

For latter like we can set noDataProgressIntervalKey to 1000000 and remove the last assertion, which isn't actually testing the behavior of deduplication. Even better, you can still keep the last assertion for only when `flag == true`.

Reply:
SPARK-31928 is filed to address the flakiness of this test on the master branch - that said, if we can make sure this test become no longer flaky, that should be nice.